keras_nlp/evaluations/glue/glue.py

```diff
@@ -0,0 +1,287 @@
+
+    # This format is used for distribution strategy compatibility.
+    def eval_step(iterator):
+        def step_fn(inputs):
+            x, _ = inputs
+            prob = finetuning_model(x)
+            pred = tf.argmax(prob, -1)
+            return pred
+
+        return step_fn(next(iterator))
+
+    labelname = labelnames.get(FLAGS.task_name)
+    test_iterator = iter(test_ds)
+    with tf.io.gfile.GFile(filename, "w") as f:
+        # GLUE requires a format of index + tab + prediction.
+        writer = csv.writer(f, delimiter="\t")
+        # Write the required headline for GLUE.
+        writer.writerow(["index", "prediction"])
+        for i in range(test_ds.cardinality()):
+            # TODO(chenmoneygithub): Add distribution strategy support.
+            pred = eval_step(test_iterator)
+            pred = pred.numpy()
+            for j in range(len(pred)):
+                idx = i * FLAGS.batch_size + j
+                if labelname:
+                    pred_value = labelname[int(pred[j])]
+                else:
+                    pred_value = pred[j]
+                writer.writerow([idx, pred_value])
+            break
+
+
+@keras.utils.register_keras_serializable(package="custom")
+class GlueClassifier(keras.Model):
+    """Default classification model for GLUE tasks.
+
+    `GlueClassifier` takes in a pretrained model as `backbone`, which must
+    output a single vector representation per input data. If the pretrained
+    model outputs a dictionary, users need to make a wrapper model to
+    extract out the representation.
+
+
+    Args:
+        backbone: A keras model instance. The output must be a single vector
+            representation per input data.
+        num_classes: int, the number of classes to predict.
+    """
+
+    def __init__(
+        self,
+        backbone,
+        num_classes=2,
+    ):
+        inputs = backbone.input
+        representation = backbone(inputs)
+        outputs = keras.layers.Dense(
+            num_classes,
+            kernel_initializer=keras.initializers.TruncatedNormal(stddev=0.02),
+            name="logits",
+        )(representation)
+        # Instantiate using Functional API Model constructor.
+        super().__init__(inputs=inputs, outputs=outputs)
+        # All references to `self` below this line.
+        self.backbone = backbone
+        self.num_classes = num_classes
+
+    def get_config(self):
+        return {
+            "backbone": keras.layers.serialize(self.backbone),
+            "num_classes": self.num_classes,
+        }
+
+    @classmethod
+    def from_config(cls, config):
+        if "backbone" in config:
+            config["backbone"] = keras.layers.deserialize(config["backbone"])
+        return cls(**config)
+
+
+def main(_):
+    train_ds, test_ds, val_ds = load_data(FLAGS.task_name)
+
+    # ----- Custom code block starts -----
+    bert_model = keras_nlp.models.Bert.from_preset("bert_tiny_uncased_en")
```

Comment:
I'm still a little confused at the overall journey we are trying to solve with this. If the goal is just to let users take this script, fork it, and add their own code, it seems simplest just not provide any of this `use_default_classifier` flag.

We can have a simple code block here.

```python
# ----- Override this section with your custom model -----
preprocessor = keras_nlp.models.BertPreprocessor.from_preset("bert_tiny_uncased_en")
backbone = keras_nlp.models.Bert.from_preset("bert_tiny_uncased_en")
classifier = keras_nlp.models.BertClassifier(
    backbone,
    num_classes=3 if FLAGS.task_name in ("mnli", "ax") else 2,
)
# ----- End section -----
```

If we do this, we have consolidated to a single section the user would be expected to override with custom model code. They can look up `BertClassifier` code to see how to do this for Bert specifically, and if they are working on their own model outside of KerasNLP, will need to figure out a classification approach that works for them (quite likely without functional model support if they are bringing their own model).

Reply:
If the goal is to avoid looking at the classifier code, why not this?

```python
# ----- Override this section with your custom model -----
preprocessor = keras_nlp.models.BertPreprocessor.from_preset("bert_tiny_uncased_en")
backbone = keras_nlp.models.Bert.from_preset("bert_tiny_uncased_en")

inputs = backbone.inputs
pooled = backbone(inputs)["pooled_output"]
outputs = keras.layers.Dense(
    num_classes,
    kernel_initializer=keras.initializers.TruncatedNormal(stddev=0.02),
)(pooled)
classifier = keras.Model(inputs, outputs)
# ----- End section -----
```

Worry is twofold...

If the main use journey is to take this and "fork" it, I don't see the point of the `use_default_classifier` argument. A flag is only useful if you don't have to touch code. But because of this setup a user will now have three different code blocks marked "change me". I think this will be too confusing of a user experience. The first think I would do if I was forking this would be to delete the `use_default_classifier` argument and choose one code path for setting up the classifier head.

The second worry is that this head is wrong not just in output alignment but in structure for every model I can think of that is not BERT. For good classifier performance, you generally want a hidden dense with a non-linearity before your output dense. BERT is an exception in that it has this hidden dense baked into the "backbone." This classifier would lead to substandard performance for distilbert, roberta, xlm-roberta, gpt-2, etc. Given that the point of GLUE is to produce a standard score for a model, if I put in a lot of work only to realize `GlueClassifier` was leading to a totally non-standard and sub-par score, that would be a frustrating experience I think.

I think you could still show the complete code for classification with a code block like above, without setting the false expectation that there is a a general `GlueClassifier` head that can lead to a correct GLUE benchmark score across a wide swath of models.

Reply:
Yea, agree that the flag is useless, I am onboarded with the idea to only keep on branch! The problem is which branch to keep, which is similar to your second worry.

For your second worry, you are right we should consider models directly returning the output of a TransformerEncoder layer. My feeling is the following 2 approaches could both work:
- Always add a dense -> activation combo to the head.
- Have an argument `has_fc_layer` to control if an extra dense layer is used. 

At a high level, my intention is "users should just read the script itself to figure out how to benchmark their own models". Based on our discussion, I think I will:
1. Delete the branch that uses `BertClassifier`.
2. Modify the `GlueClassifier` to have the nonlinear mapping. 
3. Put a disclaimer on top of `GlueClassifier` saying this is a good starting point, but you may need to make adjustments for a better result, which will also be documented in the README. 

Any concerns with the action item?

Reply:
I think that sounds like an improvement, though I'm still unconvinced that adding a classifier head here would be a value add.

Say this was our customizable code block.

```python
# ----- Override this section with your custom model -----
preprocessor = keras_nlp.models.BertPreprocessor.from_preset("bert_tiny_uncased_en")

def preprocess(x, y):
    return preprocessor(x, y)

classifier = keras_nlp.models.BertClassifier.from_preset(
    "bert_tiny_uncased_en",
    num_classes=num_classes,
)
# ----- End section -----
```

You could take that to any huggingface model, like so

```python
# ----- Override this section with your custom model -----
tokenizer = transformers.TFBertTokenizer.from_pretrained("bert-base-uncased")

def preprocess(x, y):
    return ...

classifier = transformers.TFBertForSequenceClassification.from_pretrained(
    "bert-base-uncased",
    num_labels=num_classes,
)
# ----- End section -----
```

Or tensorflow/models, like so

```python
# ----- Override this section with your custom model -----
tokenizer = ...

def preprocess(x, y):
    return ...

encoder_config = tfm.nlp.encoders.EncoderConfig(...)
encoder = tfm.nlp.encoders.build_encoder(encoder_config)
classifier = tfm.nlp.models.BertClassifier(network=encoder, num_classes=2)
# ----- End section -----
```

If you are training a model from scratch, you can write your own classifier no problem (and only you know how). If you are using a pretrained model, I don't know of any major "source" that won't also provide the classifier, handling all the model specific gotchas!

We are better off encouraging users to look for the right classifier symbol that build their own thing here (especially if they are newbies!). Even beyond the double dense, there's which token representation to pull (for roberta token 0, for gpt2 backbone the last token), and the seq2seq approach would totally blow all of this up.

Reply:
> you can write your own classifier no problem

This is actually a problem based on my experience. Without an explicit hint, this could be confusing. The problem with using `BertClassifier` is it indicates that this is made for `Bert`, not for general purpose. I don't think the token-level variation affects the classifier? Those special tokens won't be used in the classifier since we let users output the single representation. 

`seq2seq` - it's a little complex for sure, and I don't think there is a standard approach for that? T5 uses the prompt way, but I am not sure if that's how people go for now. My take is we can ignore the `seq2seq` case for now.